File: app/src/main/java/com/smedialink/abakarmagomedov/githubapplication/presentation/UsersActivity/UsersPresenterImp.java
package com.smedialink.abakarmagomedov.githubapplication.presentation.UsersActivity;

import com.smedialink.abakarmagomedov.githubapplication.data.entity.User;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by abakarmagomedov on 15/05/17.
 */

public class UsersPresenterImp implements UsersPresenter {

    private Disposable disposable;
    private final UsersInteractor interactor;
    private UsersView view;

    @Inject
    public UsersPresenterImp(UsersInteractor interactor) {
        this.interactor = interactor;
    }

    @Override
    public void getUsers() {
        disposable = interactor.loadUsersFromGit()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe(new Consumer<Disposable>() {
                    @Override
                    public void accept(Disposable disposable1) throws Exception {
                        view.showDialog();
                    }
                })
                .doOnTerminate(new Action() {
                    @Override
                    public void run() {
                        view.hideDialog();
                    }
                })
                .subscribe(new Consumer<List<User>>() {
                    @Override
                    public void accept(List<User> users) throws Exception {
                        view.showUsers(users);
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        view.error();
                    }
                });
    }

    @Override
    public void attachView(UsersView view) {
        this.view = view;
    }

    @Override
    public void detachView() {
        this.view = null;
        if(!disposable.isDisposed()){
            disposable.dispose();
        }
    }

}

File: app/src/main/java/com/smedialink/abakarmagomedov/githubapplication/presentation/UsersActivity/UsersPresenterImp.kt
package com.smedialink.abakarmagomedov.githubapplication.presentation.UsersActivity

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by abakarmagomedov on 15/05/17.
 */
class UsersPresenterImp @Inject constructor(private val interactor: UsersInteractor) :
    UsersPresenter {
    private var disposable: Disposable? = null
    private var view: UsersView? = null
    override val users: Unit
        get() {
            disposable = interactor.loadUsersFromGit()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe { view?.showDialog() }
                .doOnTerminate { view?.hideDialog() }
                .subscribe({ users -> view?.showUsers(users) }) { view?.error() }
        }

    override fun attachView(view: UsersView) {
        this.view = view
    }

    override fun detachView() {
        view = null
        if (disposable?.isDisposed == false) {
            disposable?.dispose()
        }
    }
}
